controllers/awscluster_controller_test.go

```diff
@@ -13,42 +13,1122 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+		g.Expect(err).To(BeNil())
+		setup(t)
+		awsCluster := getAWSCluster("test", ns.Name)
+
+		g.Expect(testEnv.Create(ctx, &awsCluster)).To(Succeed())
+		defer teardown()
+		defer t.Cleanup(func() {
+			g.Expect(testEnv.Cleanup(ctx, &awsCluster, controllerIdentity, ns)).To(Succeed())
+		})
+
+		cs, err := getClusterScope(awsCluster)
+		g.Expect(err).To(BeNil())
+		networkSvc := network.NewService(cs)
+		networkSvc.EC2Client = ec2Mock
+		reconciler.networkServiceFactory = func(clusterScope scope.ClusterScope) services.NetworkInterface {
+			return networkSvc
+		}
+
+		ec2Svc := ec2Service.NewService(cs)
+		ec2Svc.EC2Client = ec2Mock
+		reconciler.ec2ServiceFactory = func(scope scope.EC2Scope) services.EC2MachineInterface {
+			return ec2Svc
+		}
+		_, err = ec2Svc.EC2Client.CreateSubnet(&ec2.CreateSubnetInput{
+			VpcId:            aws.String("vpc-exists"),
+			CidrBlock:        aws.String("10.0.11.0/24"),
+			AvailabilityZone: aws.String("us-east-1c"),
+			TagSpecifications: []*ec2.TagSpecification{
+				{
+					ResourceType: aws.String("subnet"),
+					Tags: []*ec2.Tag{
+						{
+							Key:   aws.String("Name"),
+							Value: aws.String("test-cluster-subnet-public-us-east-1c"),
+						},
+						{
+							Key:   aws.String("kubernetes.io/cluster/test-cluster"),
+							Value: aws.String("shared"),
+						},
+						{
+							Key:   aws.String("kubernetes.io/role/internal-elb"),
+							Value: aws.String("1"),
+						},
+						{
+							Key:   aws.String("sigs.k8s.io/cluster-api-provider-aws/cluster/test-cluster"),
+							Value: aws.String("owned"),
+						},
+						{
+							Key:   aws.String("sigs.k8s.io/cluster-api-provider-aws/role"),
+							Value: aws.String("public"),
+						},
+					},
+				},
+			},
+		})
+		g.Expect(err).To(BeNil())
+		testSecurityGroupRoles := []infrav1.SecurityGroupRole{
+			infrav1.SecurityGroupBastion,
+			infrav1.SecurityGroupAPIServerLB,
+			infrav1.SecurityGroupLB,
+			infrav1.SecurityGroupControlPlane,
+			infrav1.SecurityGroupNode,
+		}
+		sgSvc := securitygroup.NewService(cs, testSecurityGroupRoles)
+		sgSvc.EC2Client = ec2Mock
+
+		reconciler.securityGroupFactory = func(clusterScope scope.ClusterScope) services.SecurityGroupInterface {
+			return sgSvc
+		}
+		elbSvc := elbService.NewService(cs)
+		elbSvc.EC2Client = ec2Mock
+		elbSvc.ELBClient = elbMock
+
+		reconciler.elbServiceFactory = func(elbScope scope.ELBScope) services.ELBInterface {
+			return elbSvc
+		}
+		cs.SetSubnets([]infrav1.SubnetSpec{
+			{
+				ID:               "subnet-2",
+				AvailabilityZone: "us-east-1c",
+				IsPublic:         true,
+				CidrBlock:        "10.0.11.0/24",
+			},
+			{
+				ID:               "subnet-1",
+				AvailabilityZone: "us-east-1a",
+				CidrBlock:        "10.0.10.0/24",
+				IsPublic:         false,
+			},
+		})
+		_, err = reconciler.reconcileNormal(cs)
+		g.Expect(err).To(BeNil())
+		g.Expect(cs.VPC().ID).To(Equal("vpc-exists"))
+		expectAWSClusterConditions(g, cs.AWSCluster, []conditionAssertion{
```

Comment:
Can we reload the AWSCluster object from API Server and then check the conditions? This will help in ensuring that the object was patched successfully to etcd.

Reply:
cs.AWSCluster is shows the exact conditions as it fetches the AWSCluster object from scope. ho do you suggest to reload the object?

Reply:
```
awsCluster := &infrav1.AWSCluster{}
err := r.Get(ctx, req.NamespacedName, awsCluster)
```
We can then use the assertions against this 'awsCluster' object.

Reply:
so this is the object we create, but the object gets updated in scope, I tried using this object, it contains no conditions, that's why using scope.AWSCluster

Reply:
> This will help in ensuring that the object was patched successfully to etcd
As mentioned earlier, it is advisable to run our checks on actual object present in etcd as it confirms that object is patched successfully and it is closer to the real world usecase. You can find similar approach being followed in cluster-api as well.

Reply:
not sure if the object gets persisted to etcd in these flows, as some of my tests are failing to get the conditions correctly which were fetched properly from scope

Reply:
I think it's better we leave it as is, but if you feel like we should do like this, then we have to find root cause on why no conditions are getting updated in AWSCluster object fetched from etcd

Reply:
`[this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/network/vpc.go#L80)` is one such example where we see that conditions are patched only on scope objects. Likewise, all the other services patch the conditions to scope object in same way. That's why when we try to get AWSCluster object from `testenv.Get` it doesn't contain required condition.